Compiler error when indexing dictionary on struct nested in dictionary on struct

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
OS: windows, Microsoft Windows 10 Pro v19044 64-bit
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i7-7600U CPU @ 2.80GHz  
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 66572d5

<!-- Please include the standalone minimal reproducible code.
If not, it may be closed by the administrators. -->
**What did you do?**
Line of code that breaks it:
```v
//This is the line that causes the issue:
 	val := args.nm.modules[args.thread.current_module]?.constants[name]?
    /* If instead I do this

    mod := args.nm.modules[args.thread.current_module]?
    val := mod.constants[name]?

    It works just fine
    */
```

Minimum reproducible code:
```v
import encoding.binary as bin

[heap]
struct NMachine {
	mut:
		threads shared map[u16]NThread
	pub mut:
		modules map[u64]NModule
		constants shared map[u64]NStackValue
}

pub struct NModule {
	pub mut: 
		functions map[u64]NFunction
		constants map[u64]NStackValue
}

[heap]
pub struct NFunction {
	pub mut:
		labels map[u64]int
	mut: 
		code []u8
}

[heap]
pub struct NThread {
	pub mut:
		last_index int
		current_module u64
	mut: 
		stack NStack
}

pub type NStackValue = u8 | []u8

pub struct NStack {
	mut:
		values []NStackValue
}

pub fn (mut stack NStack) push(val NStackValue) {
	stack.values << val
}

pub fn (mut stack NStack) pop() NStackValue {
	return stack.values.pop()
}

pub struct OperationArgs {
	pub: 
		opcode u8
	pub mut:
		nm &NMachine
		func &NFunction
		thread &NThread
		code &BytecodeIterator		
}


pub struct BytecodeIterator {
	arr []u8 [required]
	mut:
		idx int
}

pub fn (mut iter BytecodeIterator) next() ?u8 {
	if iter.idx >= iter.arr.len {
		return none
	}
	defer {
		iter.idx++
	}
	return iter.arr[iter.idx]
} 

pub fn (mut iter BytecodeIterator) next_u64() ?u64 {
	return read_u64(iter.next_n(8)?)
}

fn load_const(mut args OperationArgs)? {
	name := args.code.next_u64()?
    //This is the line that causes the issue:
 	val := args.nm.modules[args.thread.current_module]?.constants[name]?
    /* If instead I do this

    mod := args.nm.modules[args.thread.current_module]?
    val := mod.constants[name]?

    It works just fine
    */
	args.thread.push_stack(val)
}

pub fn read_u64(bytes []u8) u64 {
	mut nb := bytes.clone()
	len := nb.len
	if len < 8 {
		for _ in 0..(9 - len) {
			nb << u8(0)
		}
	}
	i := bin.little_endian_u64(nb)
	return i
}

pub fn (mut thread NThread) push_stack(val NStackValue) {
	thread.stack.push(val)	
}

pub fn (mut iter BytecodeIterator) next_n(n int) ?[]u8 {
	mut results := []u8{}
	for _ in 0..n {
		results << iter.next()?
	}
	return results
}

fn main(){
	println('hi')
}
```


**What did you expect to see?**
I expected my code to compile and run
 
**What did you see instead?**
```log
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:544: warning: WINVER redefined
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:943: warning: InterlockedIncrement64 redefined
In file included from C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:1279:
c:/users/myusername/source/commands/v-lang/thirdparty/tcc/include/winapi/synchapi.h:13: warning: CreateEvent redefined
In file included from C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:1279:
c:/users/myusername/source/commands/v-lang/thirdparty/tcc/include/winapi/synchapi.h:117: warning: OpenEvent redefined
In file included from C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:1279:
c:/users/myusername/source/commands/v-lang/thirdparty/tcc/include/winapi/synchapi.h:167: warning: CreateMutex redefined
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:9080: warning: implicit declaration of function 'tcc_backtrace'
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:9746: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:9746: warning: assignment makes pointer from integer without a cast
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:9746: warning: cast between pointer and integer of different size
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:15937: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:15944: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:15951: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:15959: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:15987: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:15987: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:16007: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:16009: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:16009: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:16029: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:18110: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:18201: warning: implicit declaration of function 'GetFinalPathNameByHandleW'
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:19861: warning: implicit declaration of function 'CreateSymbolicLinkW'
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:19921: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:19933: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:20396: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:20486: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22030: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22030: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22030: warning: assignment makes integer from pointer without a cast
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22100: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22100: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22100: warning: assignment makes integer from pointer without a cast
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22104: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22104: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22104: warning: assignment makes integer from pointer without a cast
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22140: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22140: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22140: warning: assignment makes integer from pointer without a cast
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22160: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22160: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22160: warning: assignment makes integer from pointer without a cast
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22181: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22181: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22181: warning: assignment makes integer from pointer without a cast
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22190: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22190: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22190: warning: assignment makes integer from pointer without a cast
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22235: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22240: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22291: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22291: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22291: warning: assignment makes integer from pointer without a cast
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22295: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22295: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22295: warning: assignment makes integer from pointer without a cast
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22361: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22366: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22386: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22386: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22386: warning: assignment makes integer from pointer without a cast
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22407: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22407: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22407: warning: assignment makes integer from pointer without a cast
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22430: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22430: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22430: warning: assignment makes integer from pointer without a cast
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22736: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22748: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22750: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22768: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22775: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22818: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22828: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22871: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:22902: warning: assignment from incompatible pointer type
C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:23529: error: lvalue expected
builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```
The c file (C:/Users/MYUSER~1/AppData/Local/Temp/v_0/my_project.4665039309180726123.tmp.c:23529) was as follows:

```c
_option_void machine__load_const(machine__OperationArgs* args) {
	_option_u64 _t1 = machine__BytecodeIterator_next_u64(args->code);
	if (_t1.state != 0) { /*or block*/ 
		_option_void _t2;
		memcpy(&_t2, &_t1, sizeof(_option));
		return _t2;
	}
	
 	u64 name =  (*(u64*)_t1.data);
	machine__NModule* _t6 = (machine__NModule*)/*ee elem_ptr_typ */(map_get_check(ADDR(map, args->nm->modules), &(u64[]){args->thread->current_module}));
	_option_machine__NModule _t5 = {0};
	if (_t6) {
		*((machine__NModule*)&_t5.data) = *((machine__NModule*)_t6);
	} else {
		_t5.state = 2; _t5.err = _v_error(_SLIT("array index out of range"));
	}
	;
	if (_t5.state != 0) { /*or block*/ 
		_option_void _t7;
		memcpy(&_t7, &_t5, sizeof(_option));
		return _t7;
	}
	
	machine__NStackValue* _t4 = (machine__NStackValue*)/*ee elem_ptr_typ */(map_get_check(ADDR(map, *(machine__NModule*)_t5.data.constants), &(u64[]){name}));
	_option_machine__NStackValue _t3 = {0};
	if (_t4) {
		*((machine__NStackValue*)&_t3.data) = *((machine__NStackValue*)_t4);
	} else {
		_t3.state = 2; _t3.err = _v_error(_SLIT("array index out of range"));
	}
	;
	if (_t3.state != 0) { /*or block*/ 
		_option_void _t8;
		memcpy(&_t8, &_t3, sizeof(_option));
		return _t8;
	}
	
	machine__NStackValue val = *(machine__NStackValue*)_t3.data;
	machine__NThread_push_stack(args->thread, val);
	return (_option_void){0};
}
```

--- a/vlib/v/gen/c/index.v
+++ b/vlib/v/gen/c/index.v
@@ -484,7 +484,7 @@ fn (mut g Gen) index_of_map(node ast.IndexExpr, sym ast.TypeSymbol) {
 			if !node.is_option {
 				g.or_block(tmp_opt, node.or_expr, elem_type)
 			}
-			g.write('\n$cur_line*($elem_type_str*)${tmp_opt}.data')
+			g.write('\n${cur_line}(*($elem_type_str*)${tmp_opt}.data)')
 		}
 	}
 }

--- a/vlib/v/tests/nested_map_index_test.v
+++ b/vlib/v/tests/nested_map_index_test.v
@@ -1,0 +1,113 @@
+import encoding.binary as bin
+
+[heap]
+struct NMachine {
+mut:
+	threads shared map[u16]NThread
+pub mut:
+	modules   map[u64]NModule
+	constants shared map[u64]NStackValue
+}
+
+pub struct NModule {
+pub mut:
+	functions map[u64]NFunction
+	constants map[u64]NStackValue
+}
+
+[heap]
+pub struct NFunction {
+pub mut:
+	labels map[u64]int
+mut:
+	code []u8
+}
+
+[heap]
+pub struct NThread {
+pub mut:
+	last_index     int
+	current_module u64
+mut:
+	stack NStack
+}
+
+pub type NStackValue = []u8 | u8
+
+pub struct NStack {
+mut:
+	values []NStackValue
+}
+
+pub fn (mut stack NStack) push(val NStackValue) {
+	stack.values << val
+}
+
+pub fn (mut stack NStack) pop() NStackValue {
+	return stack.values.pop()
+}
+
+pub struct OperationArgs {
+pub:
+	opcode u8
+pub mut:
+	nm     &NMachine
+	func   &NFunction
+	thread &NThread
+	code   &BytecodeIterator
+}
+
+pub struct BytecodeIterator {
+	arr []u8 [required]
+mut:
+	idx int
+}
+
+pub fn (mut iter BytecodeIterator) next() ?u8 {
+	if iter.idx >= iter.arr.len {
+		return none
+	}
+	defer {
+		iter.idx++
+	}
+	return iter.arr[iter.idx]
+}
+
+pub fn (mut iter BytecodeIterator) next_u64() ?u64 {
+	return read_u64(iter.next_n(8)?)
+}
+
+fn load_const(mut args OperationArgs) ? {
+	name := args.code.next_u64()?
+	val := args.nm.modules[args.thread.current_module]?.constants[name]?
+	args.thread.push_stack(val)
+}
+
+pub fn read_u64(bytes []u8) u64 {
+	mut nb := bytes.clone()
+	len := nb.len
+	if len < 8 {
+		for _ in 0 .. (9 - len) {
+			nb << u8(0)
+		}
+	}
+	i := bin.little_endian_u64(nb)
+	return i
+}
+
+pub fn (mut thread NThread) push_stack(val NStackValue) {
+	thread.stack.push(val)
+}
+
+pub fn (mut iter BytecodeIterator) next_n(n int) ?[]u8 {
+	mut results := []u8{}
+	for _ in 0 .. n {
+		results << iter.next()?
+	}
+	return results
+}
+
+fn test_nested_map_index() {
+	println('hi')
+	assert true
+}

--- a/vlib/v/tests/nested_map_index_test.v
+++ b/vlib/v/tests/nested_map_index_test.v
@@ -1,113 +1,24 @@
-import encoding.binary as bin
-
-[heap]
-struct NMachine {
+struct Foo {
 mut:
-	threads shared map[u16]NThread
-pub mut:
-	modules   map[u64]NModule
-	constants shared map[u64]NStackValue
+	foo map[int]Bar
 }
 
-pub struct NModule {
-pub mut:
-	functions map[u64]NFunction
-	constants map[u64]NStackValue
-}
-
-[heap]
-pub struct NFunction {
-pub mut:
-	labels map[u64]int
+struct Bar {
 mut:
-	code []u8
+	bar map[int]string
 }
 
-[heap]
-pub struct NThread {
-pub mut:
-	last_index     int
-	current_module u64
-mut:
-	stack NStack
-}
-
-pub type NStackValue = []u8 | u8
-
-pub struct NStack {
-mut:
-	values []NStackValue
-}
-
-pub fn (mut stack NStack) push(val NStackValue) {
-	stack.values << val
-}
-
-pub fn (mut stack NStack) pop() NStackValue {
-	return stack.values.pop()
-}
-
-pub struct OperationArgs {
-pub:
-	opcode u8
-pub mut:
-	nm     &NMachine
-	func   &NFunction
-	thread &NThread
-	code   &BytecodeIterator
-}
-
-pub struct BytecodeIterator {
-	arr []u8 [required]
-mut:
-	idx int
-}
-
-pub fn (mut iter BytecodeIterator) next() ?u8 {
-	if iter.idx >= iter.arr.len {
-		return none
-	}
-	defer {
-		iter.idx++
-	}
-	return iter.arr[iter.idx]
-}
-
-pub fn (mut iter BytecodeIterator) next_u64() ?u64 {
-	return read_u64(iter.next_n(8)?)
-}
-
-fn load_const(mut args OperationArgs) ? {
-	name := args.code.next_u64()?
-	val := args.nm.modules[args.thread.current_module]?.constants[name]?
-	args.thread.push_stack(val)
-}
-
-pub fn read_u64(bytes []u8) u64 {
-	mut nb := bytes.clone()
-	len := nb.len
-	if len < 8 {
-		for _ in 0 .. (9 - len) {
-			nb << u8(0)
+fn test_nested_map_index() ? {
+	f := Foo{
+		foo: {
+			11: Bar{
+				bar: {
+					22: 'hello'
+				}
+			}
 		}
 	}
-	i := bin.little_endian_u64(nb)
-	return i
-}
-
-pub fn (mut thread NThread) push_stack(val NStackValue) {
-	thread.stack.push(val)
-}
-
-pub fn (mut iter BytecodeIterator) next_n(n int) ?[]u8 {
-	mut results := []u8{}
-	for _ in 0 .. n {
-		results << iter.next()?
-	}
-	return results
-}
-
-fn test_nested_map_index() {
-	println('hi')
-	assert true
+	ret := f.foo[11]?.bar[22]?
+	println(ret)
+	assert ret == 'hello'
 }

